### PR TITLE
Tx validation detect duplicate inputs

### PR DESCRIPTION
**Pull request by thschroeter**

### CODE REVIEW

**Comment by thschroeter** [comment]
Review by @adagys?

Is there a better way to construct unique inputs in the tests than tweaking their amounts?

**Comment by thschroeter** [comment]
Not sure why the CI build failed. Will take another look later today.

**Comment by Clintonio** [comment]
@thschroeter that failure is a spurious one. I'll re-run it.

File: core/src/main/kotlin/net/corda/core/contracts/Structures.kt [hunk 1]
```diff
@@ -136,7 +136,7 @@ data class TransactionState<out T : ContractState> @JvmOverloads constructor(
          *
          * The encumbered state refers to another by index, and the referred encumbrance state
          * is an output state in a particular position on the same transaction that created the encumbered state. An alternative
-         * implementation would be encumber by reference to a StateRef., which would allow the specification of encumbrance
+         * implementation would be encumber by reference to a StateRef, which would allow the specification of encumbrance
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/contracts/Structures.kt]: this should probably say `encumbering by reference to a [StateRef],`

File: core/src/main/kotlin/net/corda/core/contracts/TransactionTypes.kt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package net.corda.core.contracts
 
+import kotlinx.support.jdk8.collections.forEach
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/contracts/TransactionTypes.kt]: don't think this is needed (?)

File: core/src/main/kotlin/net/corda/core/contracts/TransactionVerification.kt [hunk 3]
```diff
@@ -97,6 +97,7 @@ sealed class TransactionVerificationException(val tx: LedgerTransaction, cause:
     class SignersMissing(tx: LedgerTransaction, val missing: List<CompositeKey>) : TransactionVerificationException(tx, null) {
         override fun toString() = "Signers missing: ${missing.joinToString()}"
     }
+    class DuplicateInputStates(tx: LedgerTransaction, val duplicates: Set<StateRef>) : TransactionVerificationException(tx, null)
```
Comment (adagys) [inline core/src/main/kotlin/net/corda/core/contracts/TransactionVerification.kt]: could add a `toString()` override listing the refs

**Comment by adagys** [comment]
ok, so you shouldn't need to tweak the cash/obligation tests once https://github.com/corda/corda/pull/141 is in

**Comment by adagys** [comment]
The unique input fix is now in master

**Comment by thschroeter** [comment]
Addressed your comments, please take another look. Thanks.

**Pull request merged**